DevTools/HackStudio/LanguageServers/Cpp/ClientConnection.cpp

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2020, Itamar S. <itamar8910@gmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "ClientConnection.h"
+#include "AutoComplete.h"
+#include <AK/HashMap.h>
+#include <LibCore/File.h>
+#include <LibGUI/TextDocument.h>
+
+// #define DEBUG_CPP_LANGUAGE_SERVER
+// #define DEBUG_FILE_CONTENT
+
+namespace LanguageServers::Cpp {
+
+static HashMap<int, RefPtr<ClientConnection>> s_connections;
+
+ClientConnection::ClientConnection(NonnullRefPtr<Core::LocalSocket> socket, int client_id)
```

Comment:
Is there a reason for it to take a `NonnullRefPtr<Core::LocalSocket>` and not a `Core::LocalSocket&`?

Reply:
Well, not really but it seems to be the convention among other `IPC::ClientConnection` subclasses from what I've seen.
I mean, the constructor of `IPC::ClientConnection` itself takes a `NonnullRefPtr<Core::LocalSocket>`.

Reply:
Right...

Well, should we change that? @awesomekling?

Reply:
Hmm, IIRC there was some reason behind this. I think it was because WebContent `main()` does this:

```cpp
    auto socket = Core::LocalSocket::take_over_accepted_socket_from_system_server();
    ASSERT(socket);
    IPC::new_client_connection<WebContent::ClientConnection>(socket.release_nonnull(), 1); 
    return event_loop.exec();
```

The constructor is invoked via forwarding from `new_client_connection`.
If we take the socket by raw reference, we now hold on to it in `main()` which means we have take extra care to unref the socket, or it will leak.

Reply:
Or, not "leak" in the strict sense, but it will be forever kept alive by the stack variable in `main()`.

Reply:
Well yeah, `main()` should just unref it if it doesn't want to keep the reference. It's not hard to do, we should just use a nested scope, or am I missing something?

```cpp
    {
        auto socket = Core::LocalSocket::take_over_accepted_socket_from_system_server();
        ASSERT(socket);
        IPC::new_client_connection<WebContent::ClientConnection>(socket.release_nonnull(), 1); 
    }
    return event_loop.exec();
```